src/pymor/algorithms/bfgs.py

```diff
@@ -0,0 +1,307 @@
+                raise BFGSError
+
+        iteration += 1
+
+        # store convergence history
+        if iteration > 0 and return_stages:
+            stages.append(model.parameters.parse(mu))
+
+        active, inactive = get_active_and_inactive_sets(parameter_space, mu, eps)
+
+        # compute update to mu
+        if sum(inactive) == 0.:
+            direction = -gradient
+        else:
+            direction = project_hessian(hessian, -gradient, active, inactive)
+
+        if error_aware:
+            step_size, line_search_iteration = armijo(
+                output, mu, direction, grad=gradient, initial_value=current_output,
+                additional_criterion=error_criterion, **(line_search_params or {}))
+        else:
+            step_size, line_search_iteration = armijo(output, mu, direction, grad=gradient,
+                               initial_value=current_output, **(line_search_params or {}))
+        line_search_iterations.append(line_search_iteration + 1)
+
+        # update mu
+        old_mu = mu.copy()
+        old_output = current_output.copy()
+        mu_update = step_size * direction
+        mu += mu_update
+        mu = parameter_space.clip(mu).to_numpy()
+        current_output = output(mu)
+        data['mus'].append(mu.copy())
+
+        # compute norms
+        update_norm = np.linalg.norm(mu - old_mu)
+        update_norms.append(update_norm)
+        mu_norm = np.linalg.norm(mu)
+
+        # update gradient
+        old_gradient = gradient.copy()
+        gradient = model.parameters.parse(model.output_d_mu(mu)).to_numpy()
+        first_order_criticity = np.linalg.norm(mu - parameter_space.clip(mu - gradient).to_numpy())
+        foc_norms.append(first_order_criticity)
+
+        # set new active inactive threshhold
+        eps_update_mu = mu - gradient
+        eps = np.linalg.norm(eps_update_mu - mu)
+
+        # update relative errors
+        output_diff = abs(old_output - current_output) / abs(old_output)
+        mu_diff = np.linalg.norm(mu - old_mu) / np.linalg.norm(old_mu)
+
+        # update the hessian approximate
+        hessian = update_hessian(hessian, mu, old_mu, gradient, old_gradient)
+
+        with warnings.catch_warnings():
+            # ignore division-by-zero warnings when solution_norm or output is zero
+            warnings.filterwarnings('ignore', category=RuntimeWarning)
+            logger.info(f'it:{iteration} '
+                        f'foc:{first_order_criticity:.3e} '
+                        f'upd:{update_norm:.3e} '
+                        f'rel_upd:{update_norm / mu_norm:.3e} ')
+
+        if not np.isfinite(update_norm) or not np.isfinite(mu_norm):
+            raise BFGSError('Failed to converge.')
+
+    data['update_norms'] = np.array(update_norms)
+    data['foc_norms'] = np.array(foc_norms)
+    data['iterations'] = iteration
+    data['line_search_iterations'] = np.array(line_search_iterations)
+    if return_stages:
+        data['stages'] = np.array(stages)
```

Comment:
I think this will not work, since `stages` contains parameters. I think you can remove either `mus` or `stages` since they are the same, if I am not mistaken.